### PR TITLE
Always try to construct methods as new provider might be added

### LINKED ISSUE #18262: Provider query operation not called anymore
Since 20b6d85ab2b9cfa4cd29d2422d69c3e3f4db0a41 our internal provider tests fail.  Debugging revealed that the query operation of our provider is not called anymore.  Instead, any fetch goes to the default provider (loaded first).  This happens for RSA, EC, and DH implementations (our provider does not provide more).

### PR DESCRIPTION
**Pull request by t8m**
Fixes #18262

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
The failures are apparent. The problem is we use the method store as a kind of cache and we do not invalidate it properly.

Whenewer a new provider is loaded into a context we need to not just look into the method store, but also construct methods of all these newly added providers.

**Comment by t8m** [comment]
@levitte I'll need some help as for how to fix the problem properly.

File: crypto/core_fetch.c [hunk 1]
```diff
@@ -124,8 +124,8 @@ void *ossl_method_construct(OSSL_LIB_CTX *libctx, int operation_id,
 {
     void *method = NULL;
 
-    if ((method = mcm->get(NULL, (const OSSL_PROVIDER **)provider_rw,
-                           mcm_data)) == NULL) {
+    if (1 /*(method = mcm->get(NULL, (const OSSL_PROVIDER **)provider_rw,
+                           mcm_data)) == NULL*/) {
```
Comment (mattcaswell) [inline crypto/core_fetch.c:128]: Why is this needed?

Comment (t8m) [reply crypto/core_fetch.c:128]: Because the existing methods in the method store will match although we added a new provider and there would be a better match within that provider.

Only queries that use the preferential property match `?property=x` are affected.

Comment (mattcaswell) [reply crypto/core_fetch.c:128]: So this is a hack to get the tests to pass - but the correct solution is to fix the bug?

Comment (t8m) [reply crypto/core_fetch.c:128]: Yes, exactly. It's also in a commit marked to be dropped. Unfortunately as I said, I'll need some ideas on how to fix the bug properly.

Comment (levitte) [reply crypto/core_fetch.c:128]: Ok, so the problem seems to be that the cache isn't cleared when a new provider is loaded.  That's supposed to happen here:

https://github.com/openssl/openssl/blob/3b836385679504579ee1052ed4b4ef1d9f49fa13/crypto/provider_core.c#L1231

So the question is, why doesn't it happen.  Is `provider_flush_store_cache()` called at all?  If it is, did it fail to actually empty the query caches?

Comment (paulidale) [reply crypto/core_fetch.c:128]: I did once wonder why we needed a cache and a store...

Comment (levitte) [reply crypto/core_fetch.c:128]: > I did once wonder why we needed a cache and a store...

It's a *method* store but a *query* cache, i.e the latter saves away the result of specific queries with specific properties.

Of course, if the query cache is deemed unnecessary after all, we could throw it away.

Comment (levitte) [reply crypto/core_fetch.c:128]: > IMO no, the problem is not the cache. That is flushed properly. But the method store is the problem because in case of the optional property match the fetch from the method store will happily succeed although there would be a better match from the activated provider. So yeah, one option would be to not just flush the store cache but also the method store on provider activation. But then, why do we have a separate cache and method store after all?

Do you assume that when someone activates another provider, none of the existing methods are used?

Comment (levitte) [reply crypto/core_fetch.c:128]: I see why you're getting in trouble...  A "maybe" (`?`) property doesn't lend itself well for new providers from which nothing has been fetched before, when there are methods that match the property query with the "maybe" parts ignored.

I don't at all think it's a good idea to clear the method store when a new provider is added, because that will lead to a re-fetch of everything that was in the method store, and we just cleared a huge performance problem related to exactly that last week.

But!  It may be that this change we're discussing here is the right way to go.  After all, the construction of methods will only happen if pre-conditions allow, i.e. if methods from this particular provider aren't already in place according to the provider's opbits.  The call to `ossl_algorithm_do_all()` should therefore be relatively quick the second time around on the same input...

This does forego the *query* cache, though, but considering that its value doesn't seem to be that great, it's possible that the change on this line strikes the right balance, so to say.

I'd suggest running some performance tests on a build of this PR, just to see what difference this particular changes makes.

Comment (t8m) [reply crypto/core_fetch.c:128]: > This does forego the _query_ cache, though, but considering that its value doesn't seem to be that great, it's possible that the change on this line strikes the right balance, so to say.

It shouldn't bypass the cache - that is queried before this code is executed ossl_method_store_cache_get() is called directly from inner_evp_generic_fetch().

Comment (t8m) [reply crypto/core_fetch.c:128]: > I'd suggest running some performance tests on a build of this PR, just to see what difference this particular changes makes.

@paulidale Could you please try to run your msquic performance test environment with this PR in place? It should  hopefully not slowdown things too much.

Comment (levitte) [reply crypto/core_fetch.c:128]: > > This does forego the _query_ cache, though, but considering that its value doesn't seem to be that great, it's possible that the change on this line strikes the right balance, so to say.
> 
> It shouldn't bypass the cache - that is queried before this code is executed: ossl_method_store_cache_get() is called directly from inner_evp_generic_fetch().

Oh right...  I seem to forget that all the time ('cause yeah, when I dove into the performance problem, it took me some time to remind myself how that was done in particular)

Comment (paulidale) [reply crypto/core_fetch.c:128]: I don't have a problem emptying the store when a provider is loaded.  Loading/unloading providers is not expected to be a common operation and if it's _heavy_, it will encourage not doing it a lot.

Comment (levitte) [reply crypto/core_fetch.c:128]: > I don't have a problem emptying the store when a provider is loaded. Loading/unloading providers is not expected to be a common operation and if it's _heavy_, it will encourage not doing it a lot.

Refetching all methods is also a heavy operation.  Please let's try to avoid that

Comment (paulidale) [reply crypto/core_fetch.c:128]: There is no apparent performance loss in HPS.

Comment (t8m) [reply crypto/core_fetch.c:128]: > There is no apparent performance loss in HPS.

If that's the case, I'll cleanup the last commit so it can be actually merged.

File: crypto/core_fetch.c [hunk 2]
```diff
@@ -123,31 +123,34 @@ void *ossl_method_construct(OSSL_LIB_CTX *libctx, int operation_id,
...
-
-        /* If no method was found yet, try the global store */
-        if (method == NULL)
-            method = mcm->get(NULL, (const OSSL_PROVIDER **)provider_rw, mcm_data);
-    }
+    /*
+     * We might be tempted to try to look into the method store without
+     * constructing to see if we can find our method there already.
+     * Unfortunately that does not work well if the query contains
+     * optional properties as newly loaded providers can match them better.
```
Comment (levitte) [inline crypto/core_fetch.c:133]: I would add this:

``` text
     * We trust that ossl_method_construct_precondition() and
     * ossl_method_construct_postcondition() make sure that the
     * ossl_algorithm_do_all() does very little when methods from
     * this provider have already been constructed.
```

**Review by levitte — CHANGES_REQUESTED** [review]

File: crypto/core_fetch.c (hunk 2 above)
Comment (t8m) [reply crypto/core_fetch.c:133]: Done, except for s/this provider/a provider/

**Comment by t8m** [comment]
@mattcaswell still OK?

**Review by mattcaswell — APPROVED** [review]
Still ok

**Review by levitte — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for the reviews.
